prog/mutation.go

```diff
@@ -395,49 +458,125 @@ type mutationArgs struct {
-			return // string const
-		}
 	case *PtrType:
 		if arg.(*PointerArg).IsSpecial() {
 			// TODO: we ought to mutate this, but we don't have code for this yet.
 			return
 		}
+	default:
+		prio, ctx.Stop, canMutate = typ.getMutationPrio()
 	}
+
 	typ := arg.Type()
-	if typ == nil || typ.Dir() == DirOut || !typ.Varlen() && typ.Size() == 0 {
+	if typ == nil || typ.Dir() == DirOut || !typ.Varlen() && typ.Size() == 0 || !canMutate {
 		return
 	}
+
 	ma.args = append(ma.args, arg)
 	ma.ctxes = append(ma.ctxes, *ctx)
+	ma.priorities = append(ma.priorities, prio)
+}
+
+// Assign a priority for each type. The boolean is the reference type and it has
+// the minimum priority, since it has only two possible values.
+func (t *IntType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if (t.RangeEnd - t.RangeBegin) == 1 {
+		return minPriority, true, true
+	}
+
+	bitsCount := math.Log2((float64(t.TypeSize * 8)))
+	if t.RangeBegin == t.RangeEnd {
+		return bitsCount + 2, true, true
+	}
+
+	// an integer with a range could have a higher impact
+	return bitsCount + 4, true, true
+}
+
+func (t *StructType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return maxPriority, true, true
+}
+
+func (t *UnionType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if len(t.Fields) == 1 {
+		return minPriority, false, false
+	}
+
+	return maxPriority, true, true
+}
+
+func (t *FlagsType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if len(t.Vals) == 0 { // No value to mutate
+		return minPriority, true, false
+	}
+
+	// when bitmask is true we should a have a higher priority because we can have bitwise operations
+	if t.BitMask {
+		bitsCount := math.Log2((float64(t.TypeSize * 8)))
+		return bitsCount + 3, true, true // a higher priority than an integer without range
+	}
+
+	return 0.6 * maxPriority, true, true
+}
+
+func (t *PtrType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 3, true, true
+}
+
+func (t *ConstType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return minPriority, true, false
+}
+
+func (t *CsumType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return minPriority, true, false
+}
+
+func (t *ProcType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.5 * maxPriority, true, true
+}
+
+func (t *ResourceType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.5 * maxPriority, true, true
+}
+
+func (t *VmaType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.5 * maxPriority, true, true
+}
+
+func (t *LenType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.6 * maxPriority, true, true
+}
+
+func (t *BufferType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.8 * maxPriority, true, true // TODO: a better value for priority (maybe random in a range 0.5 - 0.8?)
```

Comment:
This `TODO` is a generic question regarding to all of the hardcoded priorities that we have here. Move it above `getMutationPrio()` definitions.